typescript/packages/plugins/ironclad/src/ironclad.service.ts

```diff
@@ -0,0 +1,735 @@
+                });
+            }
+
+            // Deposit USDC into vault
+            await walletClient.sendTransaction({
+                to: vaultAddress,
+                abi: IC_VAULT_ABI,
+                functionName: "deposit",
+                args: [parameters.tokenAmount, walletClient.getAddress()],
+            });
+
+            // Step 2: Open Trove with ic-USDC
+
+            // Check ic-USDC allowance for borrower
+            const icUSDCAllowanceResult = await walletClient.read({
+                address: vaultAddress,
+                abi: ERC20_ABI,
+                functionName: "allowance",
+                args: [walletClient.getAddress(), BORROWER_ADDRESS],
+            });
+            const icUSDCAllowance = (icUSDCAllowanceResult as { value: bigint })
+                .value;
+
+            // Approve ic-USDC if needed
+            if (Number(icUSDCAllowance) < Number(parameters.tokenAmount)) {
+                await walletClient.sendTransaction({
+                    to: vaultAddress,
+                    abi: ERC20_ABI,
+                    functionName: "approve",
+                    args: [BORROWER_ADDRESS, parameters.tokenAmount],
+                });
+            }
+
+            // Calculate hints first
+            const { upperHint, lowerHint } = await this.getHints(
+                walletClient,
+                vaultAddress,
+                BigInt(parameters.tokenAmount),
+                BigInt(parameters.iUSDAmount)
+            );
+            // Prepare openTrove parameters
+            const openTroveParams = {
+                _collateral: vaultAddress,
+                _collateralAmount: parameters.tokenAmount,
+                _maxFeePercentage: BigInt("5000000000000000"),
+                _iUSDAmount: BigInt(parameters.iUSDAmount),
+                _upperHint: upperHint as `0x${string}`,
+                _lowerHint: lowerHint as `0x${string}`,
+            };
+
+            // Execute openTrove transaction
+            const txHash = await walletClient.sendTransaction({
+                to: BORROWER_ADDRESS as `0x${string}`,
+                abi: BORROWER_ABI,
+                functionName: "openTrove",
+                args: [
+                    openTroveParams._collateral,
+                    openTroveParams._collateralAmount,
+                    openTroveParams._maxFeePercentage,
+                    openTroveParams._iUSDAmount,
+                    openTroveParams._upperHint,
+                    openTroveParams._lowerHint,
+                ],
+            });
+
+            return `Successfully deposited ${parameters.tokenAmount} USDC into ic-USDC vault and borrowed ${parameters.iUSDAmount} iUSD. Transaction: ${txHash.hash}`;
+        } catch (error) {
+            throw Error(`Failed to borrow iUSD: ${error}`);
+        }
+    }
+
+    @Tool({
+        name: "repay_iusd_ironclad",
+        description: "Repay all iUSD and close the Trove position",
+    })
+    async repayIUSD(
```

Comment:
This feels like it should be split into different steps. One is checking debt another one is withdrawing.